components/concurrency_manager/src/lib.rs

```diff
@@ -205,6 +240,45 @@ impl ConcurrencyManager {
         Ok(())
     }
 
+    // new_ts is greater than limit, or the approximate limit.
+    // To avoid false positive and guarantee TiKV availability, we need to
+    // double-check the new_ts with PD TSO.
+    fn double_check(
+        &self,
+        new_ts: TimeStamp,
+        limit: TimeStamp,
+        source: impl slog::Value + Display,
+        using_approximate: bool,
+    ) -> Result<(), crate::InvalidMaxTsUpdate> {
+        error!("possible invalid max-ts update; double checking";
+            "attempted_ts" => new_ts,
+            "max_allowed" => limit.into_inner(),
+            "source" => &source,
+            "using_approximate" => using_approximate,
+            "TSO_TIMEOUT" => ?TSO_TIMEOUT,
+        );
+        if let Some(tso) = &self.tso {
+            match block_on_timeout(tso.get_tso(), TSO_TIMEOUT) {
+                Ok(Ok(ts)) => {
+                    self.set_max_ts_limit(TimeStamp::compose(
```

Comment:
Since we already have the configuration `max_ts_drift_allowance_ms` in `ConcurrencyManager`, what do you think about this usage:

```rust
    pub fn set_max_ts_limit(&self, latest_tso: TimeStamp) {
        if latest_tso.is_max() {
            error!("max_ts_limit cannot be set to u64::max");
            return;
        }
        let limit = TimeStamp::compose(
            latest_tso.physical() + self.max_ts_drift_allowance_ms.load(Ordering::SeqCst),
            latest_tso.logical(),
        );
        ...
    }
```

Then we can change the update here to `self.set_max_ts_limit(ts)` and the update in `server.rs` to `cm.set_max_ts_limit(ts)`

https://github.com/tikv/tikv/blob/6ff4b9d4bf63dab5ffefb5be76524e6d23b26f71/components/server/src/server.rs#L1185